src/edu/stanford/nlp/mt/tm/DynamicTranslationModel.java

```diff
@@ -1104,8 +1099,10 @@ public void incrementBackward(ReorderingTypes type) {
     private final int hashCode;
     public AlignmentTemplate(SampledRule rule) {
       this.rule = rule;
-      this.hashCode = MurmurHash2.hash32(rule.f2eAll(), rule.sourceLength(), 1) ^ 
-          MurmurHash2.hash32(rule.e2fAll(), rule.targetLength(), 1);
+      String e2fString = String.join(",", org.apache.commons.lang3.ArrayUtils.toStringArray(rule.e2fAll()));
+      String f2eString = String.join(",", org.apache.commons.lang3.ArrayUtils.toStringArray(rule.f2eAll()));
```

Comment:
This doesn't seem right.  The original code is calling MurmerHash#hash32(int[], int, int), since `f2eall()` returns `int[]` (more on this below).

What we're doing is creating a string representation of the array and then hashing that string, which seems like fundamentally a different operation.

Reply:
The alignments at this stage is an `int[]` structure because of the `flatten` function. But since the alignments now are multidimensional, using the `String` parameter seemed like the best possible option

Reply:
Ahhhhhhhhhhhhhhhhhh you're right.  Ok.